include/openmc/random_ray/parallel_map.h

```diff
@@ -0,0 +1,201 @@
+    using pointer = void; // Not providing pointer semantics.
+    using reference = value_type;
+
+    iterator(std::vector<Bucket>* buckets, std::size_t bucket_index,
+      typename std::unordered_map<KeyType, std::unique_ptr<ValueType>,
+        HashFunctor>::iterator inner_it)
+      : buckets_(buckets), bucket_index_(bucket_index), inner_it_(inner_it)
+    {
+      // Advance to the first valid element if necessary.
+      advance_to_valid();
+    }
+
+    // Dereference returns a pair of (key, value).
+    reference operator*() const
+    {
+      return {inner_it_->first, *inner_it_->second};
+    }
+
+    iterator& operator++()
+    {
+      ++inner_it_;
+      advance_to_valid();
+      return *this;
+    }
+
+    iterator operator++(int)
+    {
+      iterator tmp = *this;
+      ++(*this);
+      return tmp;
+    }
+
+    bool operator==(const iterator& other) const
+    {
+      // Two iterators are equal if they refer to the same bucket vector and are
+      // both at end, or if they have the same bucket index and inner iterator.
+      return buckets_ == other.buckets_ &&
+             bucket_index_ == other.bucket_index_ &&
+             (bucket_index_ == buckets_->size() ||
+               inner_it_ == other.inner_it_);
+    }
+
+    bool operator!=(const iterator& other) const { return !(*this == other); }
+
+  private:
+    // Helper function: if we are at the end of the current bucket, advance to
+    // the next non-empty bucket.
+    void advance_to_valid()
+    {
+      while (bucket_index_ < buckets_->size() &&
+             inner_it_ == (*buckets_)[bucket_index_].map_.end()) {
+        ++bucket_index_;
+        if (bucket_index_ < buckets_->size())
+          inner_it_ = (*buckets_)[bucket_index_].map_.begin();
+      }
+    }
+
+    std::vector<Bucket>* buckets_;
+    std::size_t bucket_index_;
+    typename std::unordered_map<KeyType, std::unique_ptr<ValueType>,
+      HashFunctor>::iterator inner_it_;
+  };
+
+public:
+  //----------------------------------------------------------------------------
+  // Constructor
+  ParallelMap(int n_buckets = 1000) : buckets_(n_buckets) {}
+
+  //----------------------------------------------------------------------------
+  // Public Methods
+  void lock(const KeyType& key)
+  {
+    Bucket& bucket = get_bucket(key);
+    bucket.lock_.lock();
+  }
+
+  void unlock(const KeyType& key)
+  {
+    Bucket& bucket = get_bucket(key);
+    bucket.lock_.unlock();
+  }
+
+  void clear()
+  {
+    for (auto& bucket : buckets_) {
+      bucket.map_.clear();
+    }
+  }
+
+  bool contains(const KeyType& key)
+  {
+    Bucket& bucket = get_bucket(key);
+    // C++20
+    // return bucket.map_.contains(key);
+    return bucket.map_.find(key) != bucket.map_.end();
+  }
+
+  ValueType& operator[](const KeyType& key)
+  {
+    Bucket& bucket = get_bucket(key);
+    return *bucket.map_[key].get();
+  }
+
+  ValueType* emplace(KeyType& key, const ValueType& value)
```

Comment:
Typically the key would be taken by value or const reference

Reply:
Makes sense! I've made the update.